File: starters_example/src/test/java/com/example/config/JwtCookieAuthenticationFilterTest.java
package com.example.config;

import com.zipe.jwt.JwtTokenProvider;
import jakarta.servlet.http.Cookie;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;
import org.springframework.mock.web.MockFilterChain;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.mock.web.MockHttpServletResponse;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * {@link JwtCookieAuthenticationFilter} 的單元測試（不需 Spring context／資料庫）。
 *
 * <p>驗證混合模式下「從 cookie 讀 JWT → 驗證 → 設定 SecurityContext」的核心行為。</p>
 */
class JwtCookieAuthenticationFilterTest {

    private static final String COOKIE = "JWT_TOKEN";

    private final JwtTokenProvider tokenProvider = mock(JwtTokenProvider.class);
    private final UserDetailsService userDetailsService = mock(UserDetailsService.class);
    private final JwtCookieAuthenticationFilter filter =
            new JwtCookieAuthenticationFilter(tokenProvider, userDetailsService, COOKIE);

    @AfterEach
    void clearContext() {
        SecurityContextHolder.clearContext();
    }

    /**
     * 帶有效 token 的 cookie：應載入使用者並把已認證的 token 寫入 SecurityContext。
     */
    @Test
    void validCookieTokenSetsAuthentication() throws Exception {
        UserDetails user = User.withUsername("user01").password("x")
                .authorities(List.of(new SimpleGrantedAuthority("ORDER_EXPORT"))).build();
        when(tokenProvider.validateAndGetUsername("good-token")).thenReturn("user01");
        when(userDetailsService.loadUserByUsername("user01")).thenReturn(user);

        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setCookies(new Cookie(COOKIE, "good-token"));

        filter.doFilter(request, new MockHttpServletResponse(), new MockFilterChain());

        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        assertThat(auth).isNotNull();
        assertThat(auth.getName()).isEqualTo("user01");
        assertThat(auth.getAuthorities()).extracting("authority").containsExactly("ORDER_EXPORT");
    }

    /**
     * 無 cookie：不應建立任何認證。
     */
    @Test
    void noCookieDoesNotAuthenticate() throws Exception {
        MockHttpServletRequest request = new MockHttpServletRequest();

        filter.doFilter(request, new MockHttpServletResponse(), new MockFilterChain());

        assertThat(SecurityContextHolder.getContext().getAuthentication()).isNull();
    }

    /**
     * token 無效（驗證拋例外）：不應建立認證，且不應中斷請求鏈。
     */
    @Test
    void invalidTokenDoesNotAuthenticate() throws Exception {
        when(tokenProvider.validateAndGetUsername("bad")).thenThrow(new RuntimeException("invalid"));

        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setCookies(new Cookie(COOKIE, "bad"));

        filter.doFilter(request, new MockHttpServletResponse(), new MockFilterChain());

        assertThat(SecurityContextHolder.getContext().getAuthentication()).isNull();
    }
}

File: example-kotlin/src/test/kotlin/com/example/config/JwtCookieAuthenticationFilterTest.kt
package com.example.config

import com.zipe.jwt.JwtTokenProvider
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import jakarta.servlet.http.Cookie
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.springframework.mock.web.MockFilterChain
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.mock.web.MockHttpServletResponse
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService

/**
 * [JwtCookieAuthenticationFilter] 的單元測試（不需 Spring context／資料庫）。
 *
 * 驗證混合模式下「從 cookie 讀 JWT → 驗證 → 設定 SecurityContext」的核心行為。
 */
class JwtCookieAuthenticationFilterTest : FunSpec({

    val cookieName = "JWT_TOKEN"

    val tokenProvider = mock(JwtTokenProvider::class.java)
    val userDetailsService = mock(UserDetailsService::class.java)
    val filter = JwtCookieAuthenticationFilter(tokenProvider, userDetailsService, cookieName)

    // 每個測試後清除 SecurityContext，避免互相污染
    afterTest {
        SecurityContextHolder.clearContext()
    }

    test("帶有效 token 的 cookie：應載入使用者並把已認證的 token 寫入 SecurityContext") {
        val user: UserDetails = User.withUsername("user01").password("x")
            .authorities(listOf(SimpleGrantedAuthority("ORDER_EXPORT"))).build()
        `when`(tokenProvider.validateAndGetUsername("good-token")).thenReturn("user01")
        `when`(userDetailsService.loadUserByUsername("user01")).thenReturn(user)

        val request = MockHttpServletRequest()
        request.setCookies(Cookie(cookieName, "good-token"))

        filter.doFilter(request, MockHttpServletResponse(), MockFilterChain())

        val auth = SecurityContextHolder.getContext().authentication
        auth.shouldNotBeNull()
        auth.name shouldBe "user01"
        auth.authorities.map { it.authority } shouldContainExactly listOf("ORDER_EXPORT")
    }

    test("無 cookie：不應建立任何認證") {
        val request = MockHttpServletRequest()

        filter.doFilter(request, MockHttpServletResponse(), MockFilterChain())

        SecurityContextHolder.getContext().authentication shouldBe null
    }

    test("token 無效（驗證拋例外）：不應建立認證，且不應中斷請求鏈") {
        `when`(tokenProvider.validateAndGetUsername("bad")).thenThrow(RuntimeException("invalid"))

        val request = MockHttpServletRequest()
        request.setCookies(Cookie(cookieName, "bad"))

        filter.doFilter(request, MockHttpServletResponse(), MockFilterChain())

        SecurityContextHolder.getContext().authentication shouldBe null
    }
})
